Block.json: Refactor and stabilize selectors API

ramonjd opened the pull request:
This is a backport PR for WordPress 6.3 that includes the following PHP Gutenberg changes:

- https://github.com/WordPress/gutenberg/pull/46496
- https://github.com/WordPress/gutenberg/pull/49427
- https://github.com/WordPress/gutenberg/pull/49423 (changes not applied by the [duotone backport](https://github.com/WordPress/wordpress-develop/pull/4619))
- https://github.com/WordPress/gutenberg/pull/49393 (changes not applied by the [duotone backport](https://github.com/WordPress/wordpress-develop/pull/4619))

Related (original code introduced in 6.2):

- https://github.com/WordPress/wordpress-develop/pull/3942

Trac ticket: https://core.trac.wordpress.org/ticket/58586

cc @aaronrobertshaw

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

ramonjd committed:
Changes from https://github.com/WordPress/gutenberg/pull/49427

ramonjd commented:
@aaronrobertshaw

A pretty rough start. I wasn't sure whether the following change to blocks.php had to be ported. 

https://github.com/WordPress/gutenberg/pull/46496/files#diff-12ada35a09133c5735dcc286bb738745777c30420bc40dbc04d751db5ecfe349R21

ramonjd commented:
Didn't edit unit tests yet. Failures are expected 😄

aaronrobertshaw commented:
> A pretty rough start. I wasn't sure whether the following change to blocks.php had to be ported.

I can confirm, the changes referred to are expected to remain Gutenberg-only until 6.3 is the minimum supported version. That filter is there to ensure the selectors configuration from block.json metadata is honoured.

> Didn't edit unit tests yet. Failures are expected 😄

Most of the test failures will relate to the [removal](https://github.com/WordPress/gutenberg/pull/49393/files#diff-f5a40e066d6976361bb847177c43ccf7c1d0f5ad26555e669fa5dcf0e852c649) of the "editor-only" selectors from the API.

The theme.json tests look to be failing due to the `image/block.json` not being updated yet.

I'll have some bandwidth tomorrow to pick this up or help out. Thanks for getting this started @ramonjd 🙇

spacedmonkey commented on tests/phpunit/tests/theme/wpGetBlockCssSelector.php:
```diff
@@ -0,0 +1,445 @@
+<?php
+
+require_once __DIR__ . '/base.php';
```
What is this?

ramonjd replied:
Theme [setup scripts](https://github.com/WordPress/wordpress-develop/blob/trunk/tests/phpunit/tests/theme/base.php).

Not needed here. Removed in [e1c3210](https://github.com/WordPress/wordpress-develop/pull/4655/commits/e1c3210b3a2b4870cad3aad937b1444c739f4fc3)

ramonjd committed:
Updating classnames for wpGetBlockCssSelector.php tests
Reverting changes to Tests_Theme_wpThemeJson that will need to be reinstated after https://github.com/WordPress/wordpress-develop/pull/4619 merges

ramonjd committed:
Changes from https://github.com/WordPress/gutenberg/pull/46496

ramonjd commented:
> The theme.json tests look to be failing due to the image/block.json not being updated yet.

🤔 I think we should revert that test until the block library package is backported (done). That should happen early next week. 

> Most of the test failures will relate to the [removal](https://github.com/WordPress/gutenberg/pull/49393/files#diff-f5a40e066d6976361bb847177c43ccf7c1d0f5ad26555e669fa5dcf0e852c649) of the "editor-only" selectors from the API.

I've fixed these for now.

I noticed that https://github.com/WordPress/wordpress-develop/pull/4619 is backporting the duotone changes only, so it looks like we'll have to port over the rest of the changes from 

- https://github.com/WordPress/gutenberg/pull/49423
- https://github.com/WordPress/gutenberg/pull/49393

I've also removed the following test from `Tests_Theme_WpGetBlockCssSelector`, which was failing and I couldn't work out if it was required since the API was stabilized or because there are dependencies in https://github.com/WordPress/wordpress-develop/pull/4619. 

```php
	public function test_get_duotone_selector_via_experimental_property() {
		$block_type = self::register_test_block(
			'test/experimental-duotone-selector',
			null,
			array(
				'color' => array(
					'__experimentalDuotone' => '.experimental-duotone',
				),
			)
		);

		$selector = wp_get_block_css_selector( $block_type, 'filters.duotone' );
		$this->assertEquals( '.experimental-duotone', $selector );
	}
```

More to work to come on this one...

ramonjd committed:
Remove theme setup script (we don't need theme-based functionality for the Tests_Theme_WpGetBlockCssSelector block tests)

ramonjd committed:
Backport changes from https://github.com/WordPress/gutenberg/pull/49393

ramonjd commented:
I'm getting [1 failure](https://github.com/WordPress/wordpress-develop/actions/runs/5332856362/jobs/9662591201?pr=4655#step:15:252) for `Tests_Theme_wpThemeJson::test_get_stylesheet` where the image block class is being output twice for duotone:  `.wp-block-image .wp-block-image img, .wp-block-image .wp-block-image .components-placeholder{filter: var(--wp--preset--duotone--custom-duotone);}`. 

Is that also dependent on block.json or the duotone backport changes @aaronrobertshaw ?

aaronrobertshaw commented:
> I'm getting [1 failure](https://github.com/WordPress/wordpress-develop/actions/runs/5332856362/jobs/9662591201?pr=4655#step:15:252)

My [expectation](https://github.com/WordPress/wordpress-develop/pull/4655#issuecomment-1600336802) was that this test would be failing due to the image block.json needing to be updated. I'll go digging and see what I can find and see what else might need backporting.

aaronrobertshaw commented:
> I'm getting one failure

There's a [duotone-related change](https://github.com/WordPress/gutenberg/pull/49423/files#diff-b03597cc3da199e5aa5a94950e8241135904853e7c3b82d758e42744878afae7R2403) that we'll need to include in this backport as well to get the tests passing.

Unfortunately, I don't have permission to push to your fork or this PR branch.  Instead, below is a small diff that you can apply.

```diff
diff --git a/src/wp-includes/class-wp-theme-json.php b/src/wp-includes/class-wp-theme-json.php
index 97f6d12511..92abe30f1c 100644
--- a/src/wp-includes/class-wp-theme-json.php
+++ b/src/wp-includes/class-wp-theme-json.php
@@ -2389,8 +2389,7 @@ class WP_Theme_JSON {
 
 		// 3. Generate and append the rules that use the duotone selector.
 		if ( isset( $block_metadata['duotone'] ) && ! empty( $declarations_duotone ) ) {
-			$selector_duotone = static::scope_selector( $block_metadata['selector'], $block_metadata['duotone'] );
-			$block_rules     .= static::to_ruleset( $selector_duotone, $declarations_duotone );
+			$block_rules .= static::to_ruleset( $block_metadata['duotone'], $declarations_duotone );
 		}
 
 		// 4. Generate Layout block gap styles.

```

aaronrobertshaw reviewed (commented):
I've gone over the related Gutenberg Block Selector API PRs.

- https://github.com/WordPress/gutenberg/pull/46496
- https://github.com/WordPress/gutenberg/pull/49393
- https://github.com/WordPress/gutenberg/pull/49423
- https://github.com/WordPress/gutenberg/pull/49427

In addition to the change noted in my [earlier comment](https://github.com/WordPress/wordpress-develop/pull/4655#issuecomment-1601989611), there are a few changes to the tests that require backporting still.

A patch for the proposed updates can be found below:  
  
<details>
<summary><strong>Proposed Updates</strong></summary>  
  
```diff
diff --git a/src/wp-includes/class-wp-theme-json.php b/src/wp-includes/class-wp-theme-json.php
index 97f6d12511..92abe30f1c 100644
--- a/src/wp-includes/class-wp-theme-json.php
+++ b/src/wp-includes/class-wp-theme-json.php
@@ -2389,8 +2389,7 @@ class WP_Theme_JSON {
 
 		// 3. Generate and append the rules that use the duotone selector.
 		if ( isset( $block_metadata['duotone'] ) && ! empty( $declarations_duotone ) ) {
-			$selector_duotone = static::scope_selector( $block_metadata['selector'], $block_metadata['duotone'] );
-			$block_rules     .= static::to_ruleset( $selector_duotone, $declarations_duotone );
+			$block_rules .= static::to_ruleset( $block_metadata['duotone'], $declarations_duotone );
 		}
 
 		// 4. Generate Layout block gap styles.
diff --git a/tests/phpunit/tests/theme/wpGetBlockCssSelector.php b/tests/phpunit/tests/theme/wpGetBlockCssSelector.php
index 1a58943fc7..ac7e7629fb 100644
--- a/tests/phpunit/tests/theme/wpGetBlockCssSelector.php
+++ b/tests/phpunit/tests/theme/wpGetBlockCssSelector.php
@@ -2,6 +2,8 @@
 /**
  * Tests wp_get_block_css_selector().
  *
+ * @since 6.3.0
+ *
  * @group themes
  *
  * @covers ::wp_get_block_css_selector
@@ -18,30 +20,26 @@ class Tests_Theme_WpGetBlockCssSelector extends WP_Theme_UnitTestCase {
 	public function tear_down() {
 		unregister_block_type( $this->test_block_name );
 		$this->test_block_name = null;
-		set_current_screen( '' );
 		parent::tear_down();
 	}
 
-	private function register_test_block( $name, $selectors = null, $supports = null, $editor_selectors = null ) {
+	private function register_test_block( $name, $selectors = null, $supports = null ) {
 		$this->test_block_name = $name;
 
 		return register_block_type(
 			$this->test_block_name,
 			array(
-				'api_version'      => 2,
-				'attributes'       => array(),
-				'selectors'        => $selectors,
-				'editor_selectors' => $editor_selectors,
-				'supports'         => $supports,
+				'api_version' => 2,
+				'attributes'  => array(),
+				'selectors'   => $selectors,
+				'supports'    => $supports,
 			)
 		);
 	}
 
-	private function set_screen_to_block_editor() {
-		set_current_screen( 'edit-post' );
-		get_current_screen()->is_block_editor( true );
-	}
-
+	/**
+	 * @ticket 58586
+	 */
 	public function test_get_root_selector_via_selectors_api() {
 		$block_type = self::register_test_block(
 			'test/block-with-selectors',
@@ -52,6 +50,9 @@ class Tests_Theme_WpGetBlockCssSelector extends WP_Theme_UnitTestCase {
 		$this->assertEquals( '.wp-custom-block-class', $selector );
 	}
 
+	/**
+	 * @ticket 58586
+	 */
 	public function test_get_root_selector_via_experimental_property() {
 		$block_type = self::register_test_block(
 			'test/block-without-selectors',
@@ -63,6 +64,9 @@ class Tests_Theme_WpGetBlockCssSelector extends WP_Theme_UnitTestCase {
 		$this->assertEquals( '.experimental-selector', $selector );
 	}
 
+	/**
+	 * @ticket 58586
+	 */
 	public function test_default_root_selector_generation_for_core_block() {
 		$block_type = self::register_test_block(
 			'core/without-selectors-or-supports',
@@ -74,6 +78,9 @@ class Tests_Theme_WpGetBlockCssSelector extends WP_Theme_UnitTestCase {
 		$this->assertEquals( '.wp-block-without-selectors-or-supports', $selector );
 	}
 
+	/**
+	 * @ticket 58586
+	 */
 	public function test_default_root_selector_generation() {
 		$block_type = self::register_test_block(
 			'test/without-selectors-or-supports',
@@ -85,30 +92,9 @@ class Tests_Theme_WpGetBlockCssSelector extends WP_Theme_UnitTestCase {
 		$this->assertEquals( '.wp-block-test-without-selectors-or-supports', $selector );
 	}
 
-	public function test_get_duotone_selector_via_selectors_api() {
-		$block_type = self::register_test_block(
-			'test/duotone-selector',
-			array(
-				'filters' => array( 'duotone' => '.duotone-selector' ),
-			),
-			null
-		);
-
-		$selector = wp_get_block_css_selector( $block_type, array( 'filters', 'duotone' ) );
-		$this->assertEquals( '.duotone-selector', $selector );
-	}
-
-	public function test_no_duotone_selector_set() {
-		$block_type = self::register_test_block(
-			'test/null-duotone-selector',
-			null,
-			null
-		);
-
-		$selector = wp_get_block_css_selector( $block_type, 'filters.duotone' );
-		$this->assertEquals( null, $selector );
-	}
-
+	/**
+	 * @ticket 58586
+	 */
 	public function test_get_feature_selector_via_selectors_api() {
 		$block_type = self::register_test_block(
 			'test/feature-selector',
@@ -120,6 +106,9 @@ class Tests_Theme_WpGetBlockCssSelector extends WP_Theme_UnitTestCase {
 		$this->assertEquals( '.typography', $selector );
 	}
 
+	/**
+	 * @ticket 58586
+	 */
 	public function test_get_feature_selector_via_selectors_api_shorthand_property() {
 		$block_type = self::register_test_block(
 			'test/shorthand-feature-selector',
@@ -131,6 +120,9 @@ class Tests_Theme_WpGetBlockCssSelector extends WP_Theme_UnitTestCase {
 		$this->assertEquals( '.typography', $selector );
 	}
 
+	/**
+	 * @ticket 58586
+	 */
 	public function test_no_feature_level_selector_via_selectors_api() {
 		$block_type = self::register_test_block(
 			'test/null-feature-selector',
@@ -142,6 +134,9 @@ class Tests_Theme_WpGetBlockCssSelector extends WP_Theme_UnitTestCase {
 		$this->assertEquals( null, $selector );
 	}
 
+	/**
+	 * @ticket 58586
+	 */
 	public function test_fallback_feature_level_selector_via_selectors_api_to_generated_class() {
 		$block_type = self::register_test_block(
 			'test/fallback-feature-selector',
@@ -153,7 +148,9 @@ class Tests_Theme_WpGetBlockCssSelector extends WP_Theme_UnitTestCase {
 		$this->assertEquals( '.wp-block-test-fallback-feature-selector', $selector );
 	}
 
-
+	/**
+	 * @ticket 58586
+	 */
 	public function test_fallback_feature_level_selector_via_selectors_api() {
 		$block_type = self::register_test_block(
 			'test/fallback-feature-selector',
@@ -165,6 +162,9 @@ class Tests_Theme_WpGetBlockCssSelector extends WP_Theme_UnitTestCase {
 		$this->assertEquals( '.fallback-root-selector', $selector );
 	}
 
+	/**
+	 * @ticket 58586
+	 */
 	public function test_get_feature_selector_via_experimental_property() {
 		$block_type = self::register_test_block(
 			'test/experimental-feature-selector',
@@ -180,6 +180,9 @@ class Tests_Theme_WpGetBlockCssSelector extends WP_Theme_UnitTestCase {
 		$this->assertEquals( '.wp-block-test-experimental-feature-selector .experimental-typography', $selector );
 	}
 
+	/**
+	 * @ticket 58586
+	 */
 	public function test_fallback_feature_selector_via_experimental_property() {
 		$block_type = self::register_test_block(
 			'test/fallback-feature-selector',
@@ -191,6 +194,9 @@ class Tests_Theme_WpGetBlockCssSelector extends WP_Theme_UnitTestCase {
 		$this->assertEquals( '.wp-block-test-fallback-feature-selector', $selector );
 	}
 
+	/**
+	 * @ticket 58586
+	 */
 	public function test_no_feature_selector_via_experimental_property() {
 		$block_type = self::register_test_block(
 			'test/null-experimental-feature-selector',
@@ -202,6 +208,9 @@ class Tests_Theme_WpGetBlockCssSelector extends WP_Theme_UnitTestCase {
 		$this->assertEquals( null, $selector );
 	}
 
+	/**
+	 * @ticket 58586
+	 */
 	public function test_get_subfeature_selector_via_selectors_api() {
 		$block_type = self::register_test_block(
 			'test/subfeature-selector',
@@ -221,6 +230,9 @@ class Tests_Theme_WpGetBlockCssSelector extends WP_Theme_UnitTestCase {
 		$this->assertEquals( '.root .typography .text-decoration', $selector );
 	}
 
+	/**
+	 * @ticket 58586
+	 */
 	public function test_fallback_subfeature_selector_via_selectors_api() {
 		$block_type = self::register_test_block(
 			'test/subfeature-selector',
@@ -239,6 +251,9 @@ class Tests_Theme_WpGetBlockCssSelector extends WP_Theme_UnitTestCase {
 		$this->assertEquals( '.root .typography', $selector );
 	}
 
+	/**
+	 * @ticket 58586
+	 */
 	public function test_no_subfeature_level_selector_via_selectors_api() {
 		$block_type = self::register_test_block(
 			'test/null-subfeature-selector',
@@ -250,6 +265,9 @@ class Tests_Theme_WpGetBlockCssSelector extends WP_Theme_UnitTestCase {
 		$this->assertEquals( null, $selector );
 	}
 
+	/**
+	 * @ticket 58586
+	 */
 	public function test_fallback_subfeature_selector_via_experimental_property() {
 		$block_type = self::register_test_block(
 			'test/fallback-subfeature-selector',
@@ -265,6 +283,9 @@ class Tests_Theme_WpGetBlockCssSelector extends WP_Theme_UnitTestCase {
 		$this->assertEquals( '.wp-block-test-fallback-subfeature-selector', $selector );
 	}
 
+	/**
+	 * @ticket 58586
+	 */
 	public function test_no_subfeature_selector_via_experimental_property() {
 		$block_type = self::register_test_block(
 			'test/null-experimental-subfeature-selector',
@@ -279,6 +300,9 @@ class Tests_Theme_WpGetBlockCssSelector extends WP_Theme_UnitTestCase {
 		$this->assertEquals( null, $selector );
 	}
 
+	/**
+	 * @ticket 58586
+	 */
 	public function test_empty_target_returns_null() {
 		$block_type = self::register_test_block(
 			'test/null-experimental-subfeature-selector',
@@ -293,6 +317,9 @@ class Tests_Theme_WpGetBlockCssSelector extends WP_Theme_UnitTestCase {
 		$this->assertEquals( null, $selector );
 	}
 
+	/**
+	 * @ticket 58586
+	 */
 	public function test_string_targets_for_features() {
 		$block_type = self::register_test_block(
 			'test/target-types-for-features',
@@ -307,6 +334,9 @@ class Tests_Theme_WpGetBlockCssSelector extends WP_Theme_UnitTestCase {
 		$this->assertEquals( '.found', $selector );
 	}
 
+	/**
+	 * @ticket 58586
+	 */
 	public function test_string_targets_for_subfeatures() {
 		$block_type = self::register_test_block(
 			'test/target-types-for-features',
@@ -322,107 +352,5 @@ class Tests_Theme_WpGetBlockCssSelector extends WP_Theme_UnitTestCase {
 		$selector = wp_get_block_css_selector( $block_type, array( 'typography', 'fontSize' ) );
 		$this->assertEquals( '.found', $selector );
 	}
-
-	public function test_editor_only_root_selector() {
-		self::set_screen_to_block_editor();
-
-		$block_type = self::register_test_block(
-			'test/editor-only-selectors',
-			array( 'root' => '.wp-custom-block-class' ),
-			null,
-			array( 'root' => '.wp-custom-block-class' )
-		);
-
-		$selector = wp_get_block_css_selector( $block_type, 'root' );
-		$this->assertEquals( '.wp-custom-block-class', $selector );
-	}
-
-	public function test_editor_only_duotone_selector() {
-		self::set_screen_to_block_editor();
-
-		$block_type = self::register_test_block(
-			'test/editor-duotone-selector',
-			array(
-				'filters' => array( 'duotone' => '.duotone-selector' ),
-			),
-			null,
-			array(
-				'filters' => array( 'duotone' => '.duotone-selector' ),
-			)
-		);
-
-		$selector = wp_get_block_css_selector( $block_type, 'filters.duotone' );
-		$this->assertEquals( '.duotone-selector', $selector );
-	}
-
-	public function test_editor_only_feature_selector() {
-		self::set_screen_to_block_editor();
-
-		$block_type = self::register_test_block(
-			'test/editor-feature-selector',
-			array( 'typography' => array( 'root' => '.typography' ) ),
-			null,
-			array( 'typography' => array( 'root' => '.typography' ) )
-		);
-
-		$selector = wp_get_block_css_selector( $block_type, 'typography' );
-		$this->assertEquals( '.typography', $selector );
-	}
-
-	public function test_editor_only_feature_selector_shorthand() {
-		self::set_screen_to_block_editor();
-
-		$block_type = self::register_test_block(
-			'test/editor-feature-selector',
-			array( 'typography' => '.typography' ),
-			null,
-			array( 'typography' => '.typography' )
-		);
-
-		$selector = wp_get_block_css_selector( $block_type, 'typography' );
-		$this->assertEquals( '.typography', $selector );
-	}
-
-	public function test_editor_only_subfeature_selector() {
-		self::set_screen_to_block_editor();
-
-		$block_type = self::register_test_block(
-			'test/editor-subfeature-selector',
-			array( 'typography' => array( 'fontSize' => '.font-size' ) ),
-			null,
-			array( 'typography' => array( 'fontSize' => '.font-size' ) )
-		);
-
-		$selector = wp_get_block_css_selector( $block_type, 'typography.fontSize' );
-		$this->assertEquals( '.font-size', $selector );
-	}
-
-	public function test_non_editor_subfeature_does_not_fall_back_to_editor_only_feature_selector() {
-		self::set_screen_to_block_editor();
-
-		$block_type = self::register_test_block(
-			'test/editor-subfeature-selector',
-			array( 'typography' => array( 'fontSize' => '.font-size' ) ),
-			null,
-			array( 'typography' => '.font-size' )
-		);
-
-		$selector = wp_get_block_css_selector( $block_type, 'typography.fontSize', true );
-		$this->assertEquals( '.font-size', $selector );
-	}
-
-	public function test_unspecified_subfeature_falls_back_to_editor_only_feature_selector() {
-		self::set_screen_to_block_editor();
-
-		$block_type = self::register_test_block(
-			'test/editor-subfeature-selector',
-			array( 'typography' => '.typography' ),
-			null,
-			array( 'typography' => '.typography' )
-		);
-
-		$selector = wp_get_block_css_selector( $block_type, 'typography.fontSize', true );
-		$this->assertEquals( '.typography', $selector );
-	}
 }
 

```  

</details>

The following is a breakdown per-commit.

https://github.com/WordPress/gutenberg/pull/46496

✅ `class-wp-theme-json-gutenberg.php` - Backported
🚫 `blocks.php `- N/A - Changes are plugin-only
🚫 `lib/load.php `- N/A
❓️ `phpunit/class-wp-theme-json-test.php `- New tests haven't been backported yet although they require registered blocks using both the old experimental selector and the new selectors API. This block registration was handled in Gutenberg via the phpunit bootstrap. I’m not sure how best to approach it here for core though. 
✅ `phpunit/class-wp-get-block-css-selectors-test.php` - Backported (updates in later PRs haven't been applied to backport though).

https://github.com/WordPress/gutenberg/pull/49393

🚫 `class-wp-duotone-gutenberg.php `- N/A - To be backported in duotone-specific PR (see h[ttps://github.com/WordPress/wordpress-develop/pull/4619)](https://github.com/WordPress/wordpress-develop/pull/4619)
✅ `class-wp-theme-json-gutenberg.php` - Backported
🚫 `blocks.php `- N/A - The filter added here should be plugin only
✅ `get-global-styles-and-settings.php` - Backported
❓️ `phpunit/class-wp-get-block-css-selectors-test.php `- We need to backport the changes here to remove the editor-only settings tests and update the key duotone selectors are found under.
✅ `phpunit/class-wp-theme-json-test.php` - Already backported elsewhere

https://github.com/WordPress/gutenberg/pull/49423

🚫 `class-wp-duotone-gutenberg.php `- N/A - To be backported in duotone-specific PR (see h[ttps://github.com/WordPress/wordpress-develop/pull/4619)](https://github.com/WordPress/wordpress-develop/pull/4619)
✅ `class-wp-theme-json-gutenberg.php` - Backported
✅ `get-global-styles-and-settings.php` - Backported
❓️ `phpunit/class-wp-get-block-css-selectors-test.php `- The duotone test removals in this PR aren't reflected in the backport
✅ `phpunit/class-wp-theme-json-test.php` - Typo correction backported

https://github.com/WordPress/gutenberg/pull/49427

✅ `block-support/settings.php`  - Backported in full

ramonjd commented:
Thanks for doing the heavy thinking here @aaronrobertshaw I'll get these in tomorrow 👍

ramonjd committed:
Updating according to feedback in https://github.com/WordPress/wordpress-develop/pull/4655#pullrequestreview-1492670466

aaronrobertshaw reviewed (commented):
Thank you for wrangling those updates @ramonjd 🙇 

I've given this PR another once over and it covers all the updates from the relevant block selector PRs and is consistent with the current state of Gutenberg trunk.

The failing e2es look unrelated to me. So code wise I think this is pretty close. It would be good to get final approval from someone more familiar with applying core standards though.

ramonjd commented:
@MaggieCabrera @aaronrobertshaw 

Just checking, as there might be some overlap between this PR and 

- https://github.com/WordPress/wordpress-develop/pull/4619

Suggesting that one (maybe this PR) gets merged first and then we rebase https://github.com/WordPress/wordpress-develop/pull/4619?

MaggieCabrera commented:
> @MaggieCabrera @aaronrobertshaw
> 
> Just checking, as there might be some overlap between this PR and
> 
>     * [Backport: Duotone changes #4619](https://github.com/WordPress/wordpress-develop/pull/4619)
> 
> 
> Suggesting that one (maybe this PR) gets merged first and then we rebase #4619?

I think that's the right order, yes

aaronrobertshaw commented:
> Suggesting that one (maybe this PR) gets merged first and then we rebase https://github.com/WordPress/wordpress-develop/pull/4619?

Merging this PR first works for me, thanks 🙂 

I'm happy to help out if I can on the duotone updates as well.

ramonjd commented:
> I'm happy to help out if I can on the duotone updates as well.

That'd be great, thank you!

I think this one seems to have fallen through the cracks? https://github.com/WordPress/gutenberg/pull/50678 Not sure.

aaronrobertshaw commented:
> I think this one seems to have fallen through the cracks? https://github.com/WordPress/gutenberg/pull/50678 Not sure.

Thanks for flagging that @ramonjd. It does appear to have been missed. It is related to duotone filters rather than the block selectors API so I think belongs in https://github.com/WordPress/wordpress-develop/pull/4619.

cc/ @MaggieCabrera

aaronrobertshaw reviewed (approved):
This PR covers all the necessary backports for the block selectors API. Anything related to duotone changes will be included in https://github.com/WordPress/wordpress-develop/pull/4619.

I'm happy to give this an approval so it can be moved forward.

Thanks again for wrangling this @ramonjd 👍

tellthemachines committed:
Fix multiline comment.

tellthemachines commented:
committed in [r56058](https://core.trac.wordpress.org/changeset/56058) / [8b9d8d8](https://github.com/WordPress/wordpress-develop/commit/8b9d8d8a9f96e7e810f30e94fba3040c3ebbf61a).
